run.sh

```diff
@@ -1,15 +1,71 @@
-  devpi use http://localhost:3141
-  devpi login root --password=''
-  devpi user -m root password="${DEVPI_PASSWORD}"
+
+
+#===========================
+# Default environment if none 
+# is provided
+#===========================
+DEFAULT_DEVPI_SERVERDIR=/devpidata
+DEFAULT_NGINX_IP=0.0.0.0
+DEFAULT_NGINX_PORT=3141
+DEFAULT_DEVPI_SERVERIP=127.0.0.1
+DEFAULT_DEVPI_SERVERPORT=3142
+DEFAULT_DEVPI_CLIENTDIR=/tmp/devpi-client
+DEFAULT_DEVPI_CLIENTNAME=root
+DEFAULT_DEVPI_CLIENTPWD=
+DEFAULT_DEVPI_LDAPCONFILE=
+
+# http://stackoverflow.com/questions/228544/how-to-tell-if-a-string-is-not-defined-in-a-bash-shell-script
+[[ -z ${DEVPI_SERVERDIR+xxx} ]]   && export DEVPI_SERVERDIR=$DEFAULT_DEVPI_SERVERDIR
+[[ -z ${NGINX_IP+xxx} ]]          && export NGINX_IP=$DEFAULT_NGINX_IP
+[[ -z ${NGINX_PORT+xxx} ]]        && export NGINX_PORT=$DEFAULT_NGINX_PORT
+[[ -z ${DEVPI_SERVERIP+xxx} ]]    && export DEVPI_SERVERIP=$DEFAULT_DEVPI_SERVERIP
+[[ -z ${DEVPI_SERVERPORT+xxx} ]]  && export DEVPI_SERVERPORT=$DEFAULT_DEVPI_SERVERPORT
+[[ -z ${DEVPI_CLIENTDIR+xxx} ]]   && export DEVPI_CLIENTDIR=$DEFAULT_DEVPI_CLIENTDIR
+[[ -z ${DEVPI_CLIENTNAME+xxx} ]]  && export DEVPI_CLIENTNAME=$DEFAULT_DEVPI_CLIENTNAME
+[[ -z ${DEVPI_CLIENTPWD+xxx} ]]   && export DEVPI_CLIENTPWD=$DEFAULT_DEVPI_CLIENTPWD
+[[ -z ${DEVPI_LDAPCONFILE+xxx} ]] && export DEVPI_LDAPCONFILE=$DEFAULT_DEVPI_LDAPCONFILE
+
+#===========================
+# devpi configuration
+# TODO add ldap stuff
+#===========================
+
+# Start the server
+devpi-server --start --host ${DEVPI_SERVERIP} --port ${DEVPI_SERVERPORT}
+
+# Configure a default client if necessary
+if [[ ! -f ${DEVPI_SERVERDIR}/.serverversion ]]; then
+  devpi use http://${DEVPI_SERVERIP}:${DEVPI_SERVERPORT}
+  devpi login ${DEVPI_CLIENTNAME} --password="${DEVPI_CLIENTPWD}"
+  devpi user -m root password="${DEVPI_CLIENTPWD}"
   devpi index -y -c public pypi_whitelist='*'
 fi
-tail -f $DEVPI_SERVERDIR/.xproc/devpi-server/xprocess.log
+
+#===========================
+# log configuration
+# FIXME: devpi-server and nginx logs are mixed
+#===========================
+
+# forward request and error logs to docker log collector
+ln -sf /dev/stdout ${DEVPI_SERVERDIR}/.xproc/devpi-server/xprocess.log
+ln -sf /dev/stdout /var/log/nginx/access.log
+ln -sf /dev/stderr /var/log/nginx/error.log
+
+#===========================
+# nginx configuration
+#===========================
+
+# Edit virtual host based on the environment
+nginx_conf_file=/etc/nginx/sites-available/devpi.conf
+sed -i "s/{{NGINX_IP}}/${NGINX_IP}/g"                 $nginx_conf_file 
+sed -i "s/{{NGINX_PORT}}/${NGINX_PORT}/g"             $nginx_conf_file 
+sed -i "s/{{DEVPI_SERVERPORT}}/${DEVPI_SERVERPORT}/g" $nginx_conf_file 
+# DEVPI_SERVERDIR contains / so it needs to be escaped
+# See http://stackoverflow.com/questions/407523/escape-a-string-for-a-sed-replace-pattern
+sed -i "s/{{DEVPI_SERVERDIR}}/$(echo ${DEVPI_SERVERDIR} | sed -e 's/[\/&]/\\&/g')/g" $nginx_conf_file 
```

Comment:
I have used nginx as reverse proxy for other services, and found more practical to ship the whole nginx.conf with `http` and `server` sections in a single file than reusing default configuration files and _/etc/nginx/sites-enable_ method.

for example this is the reverse proxy for docker registry:
```
daemon off;    
user www-data;    
worker_processes 4;    
pid /run/nginx.pid;    
    
events {    
  worker_connections 768;    
}    
    
http {    
  tcp_nopush on;    
  tcp_nodelay on;    
  keepalive_timeout 65;    
  types_hash_max_size 2048;    
  access_log /var/log/nginx/access.log;    
  error_log /var/log/nginx/error.log;    
  gzip off;    
    
  upstream docker-registry {    
    server ${REGISTRY_PORT_5000_TCP_ADDR}:${REGISTRY_PORT_5000_TCP_PORT};    
  }    
    
  server {    
    listen 443;    
    server_name i.mydomain.com;    
    ssl on;    
    ssl_certificate /opt/nginx/cert.crt;    
    ssl_certificate_key /opt/nginx/cert.key;    
    
    # disable any limits to avoid HTTP 413 for large image uploads    
    client_max_body_size 0;    
    
    proxy_set_header Host           $host;    
    proxy_set_header X-Real-IP      $remote_addr;    
    proxy_set_header Authorization  "";    
    
    location /ca.crt {    
        alias /opt/nginx/ca.crt;    
    }    
    
    location / {    
      auth_basic "Registry";    
      auth_basic_user_file /opt/nginx/htpasswd;    
      proxy_pass  http://docker-registry;    
    }    
    
    location /v1/_ping {    
      proxy_pass  http://docker-registry;    
    }    
    
  }    
}    
```

I store it as nginx-tpl.conf and use a shell script to generate the actual nginx.conf file:

```bash
#!/bin/bash    
env \    
  |sed 's/[\%]/\\&/g;s/\([^=]*\)=\(.*\)/s%${\1}%\2%/' \    
  |sed -f- /opt/nginx/nginx.conf \    
  |tee /tmp/nginx.conf    
exec nginx -c /tmp/nginx.conf "$@"
```

as you see (or not) it is generating sed rules for every environment variable and using it to render the nginx conf template. Above script is docker image command.